RecoHGCal/TICL/plugins/PatternRecognitionbyKF.cc

```diff
@@ -0,0 +1,481 @@
+                                                                const std::vector<HGCDiskGeomDet *> &vec, 
+                                                                bool isSilicon) const{                                                           
+  auto it = std::find(vec.begin(), vec.end(),from);
+  if (it == vec.end()) throw cms::Exception("LogicError", "nextDisk called with invalid starting disk");
+  isSilicon? --it: ++it;
+  return *(it);
+}
+
+template<typename TILES>
+const HGCDiskGeomDet * PatternRecognitionbyKF<TILES>::nextDisk(const HGCDiskGeomDet * from, 
+                                                              PropagationDirection direction, 
+                                                              const std::vector<HGCDiskGeomDet *> &vec, 
+                                                              bool isSilicon) const{
+  auto it = std::find(vec.begin(), vec.end(),from);
+  if (it == vec.end()) throw cms::Exception("LogicError", "nextDisk called with invalid starting disk");
+  int currentLayer = (*it)->layer();
+  if (direction == alongMomentum){
+    if ((*it == vec.back()) || (*it == vec.rbegin()[1])) return nullptr; // if disk last silicon OR last scintillator disk
+    while ((*it)->layer()<currentLayer+2){
+      if ((*it) == vec.back()) break;
+      ++it;
+      if(((*(it))->isSilicon() == isSilicon) && ((*(it))->layer()==currentLayer+1)) return *(it);
+     }
+  } else{
+    if (it == vec.begin()) return nullptr;
+    while ((*it)->layer()>currentLayer-2){
+      if ((it) == vec.begin()) break;
+      --it;
+      if(((*(it))->isSilicon() == isSilicon) && ((*(it))->layer()==currentLayer-1)) return *(it);
+    }
+  }
+  return nullptr; // Returns nullptr if nextDisk reached end of detector
+}
+
+template<typename TILES>
+template<class Start> 
+std::vector<TempTrajectory>
+PatternRecognitionbyKF<TILES>::advanceOneLayer(const Start &start, 
+                                              const HGCDiskGeomDet * disk, 
+                                              const std::vector<HGCDiskGeomDet *> &disks, 
+                                              const TILES &tiles, PropagationDirection direction, 
+                                              bool &isSilicon, 
+                                              TempTrajectory traj){
+
+  std::vector<TempTrajectory> ret;
+
+  const Propagator &prop = (direction == alongMomentum ? *propagator_ : *propagatorOppo_);
+  TrajectoryStateOnSurface tsos = prop.propagate(start, disk->surface());
+  if (!tsos.isValid()) return ret;
+  float r = sqrt(pow(tsos.globalPosition().x(),2)+pow(tsos.globalPosition().y(),2));
+  if (((disk->rmin() > r) && (!isSilicon)) || (((r > disk->rmax()) && (isSilicon)))) {
+    disk = switchDisk(disk, disks, isSilicon);
+    isSilicon = !isSilicon;
+    tsos = prop.propagate(start, disk->surface());
+  }
+
+  // Collect hits with estimate
+  int depth = disk->layer()+1;
```

Comment:
I would call it `layer` if this is indeed what it is. `depth` can be a bit mysterious